### PR TITLE
Fix shared return jump to relocs

### LINKED ISSUE #202: Mishandled tailjump with reloc inside the jump instruction
```
[i] ℤ rizin numabasic.c-gcc-x64-O2.o
Warning: run rizin with -e io.cache=true to fix relocations in disassembly
 -- The '?' command can be used to evaluate math expressions. Like this: '? (0x34+22)*4'
[0x08000040]> aaa
[x] Analyze all flags starting with sym. and entry0 (aa)
[x] Analyze function calls (aac)
[x] Analyze len bytes of instructions for references (aar)
[x] Check for vtables
[x] Type matching analysis for all functions (aaft)
[x] Propagate noreturn information
[x] Use -AA or aaaa to perform additional experimental analysis.
[0x08000040]> s sym.numaWriteMem
[0x08001ac0]> pdf
╭ sym.numaWriteMem (int64_t arg1, int64_t arg2, int64_t arg3);
│           ; arg int64_t arg1 @ rdi
│           ; arg int64_t arg2 @ rsi
│           ; arg int64_t arg3 @ rdx
│           0x08001ac0      push  r13
│           0x08001ac2      push  r12
│           0x08001ac4      push  rbp
│           0x08001ac5      push  rbx
│           0x08001ac6      sub   rsp, 8
│           0x08001aca      test  rdi, rdi                             ; arg1
│           0x08001acd      mov   eax, dword [0x08001ad3]              ; [0x8001ad3:4]=0x9f840f; RELOC 32 LeptMsgSeverity
│           ; DATA XREF from sym.numaWriteMem @ 0x8001acd
│       ╭─< 0x08001ad3      je    0x8001b78
│       │   0x08001ad9      test  rsi, rsi                             ; arg2
│       │   0x08001adc      mov   qword [rdi], 0                       ; arg1
│      ╭──< 0x08001ae3      je    0x8001b58
│      ││   0x08001ae5      test  rdx, rdx                             ; arg3
│      ││   0x08001ae8      mov   qword [rsi], 0                       ; arg2
│     ╭───< 0x08001aef      je    0x8001be8
│     │││   0x08001af5      cmp   eax, 3                               ; 3
│     │││   0x08001af8      mov   r12, rdx                             ; arg3
│     │││   0x08001afb      mov   rbx, rsi                             ; arg2
│     │││   0x08001afe      mov   rbp, rdi                             ; arg1
│    ╭────< 0x08001b01      jle   stderr
│    ││││   ; CODE XREF from sym.numaWriteMem @ 0x8001c31
│   ╭─────> 0x08001b07      call  tmpfile                              ; RELOC 32 tmpfile
│   ╎││││   ; CALL XREF from sym.numaWriteMem @ 0x8001b07
│   ╎││││   0x08001b0c      test  rax, rax
│   ╎││││   0x08001b0f      mov   r13, rax
│  ╭──────< 0x08001b12      je    LeptMsgSeverity
│  │╎││││   0x08001b18      mov   rsi, r12
│  │╎││││   0x08001b1b      mov   rdi, rax
│  │╎││││   0x08001b1e      call  numaWriteStream                      ; RELOC 32 numaWriteStream @ 0x080017a0 - 0x8001b23
│  │╎││││   ; CALL XREF from sym.numaWriteMem @ 0x8001b1e
│  │╎││││   0x08001b23      mov   rdi, r13
│  │╎││││   0x08001b26      mov   r12d, eax
│  │╎││││   0x08001b29      call  rewind                               ; RELOC 32 rewind
│  │╎││││   ; CALL XREF from sym.numaWriteMem @ 0x8001b29
│  │╎││││   0x08001b2e      mov   rsi, rbx
│  │╎││││   0x08001b31      mov   rdi, r13
│  │╎││││   0x08001b34      call  l_binaryReadStream                   ; RELOC 32 l_binaryReadStream
│  │╎││││   ; CALL XREF from sym.numaWriteMem @ 0x8001b34
│  │╎││││   0x08001b39      mov   rdi, r13
│  │╎││││   0x08001b3c      mov   qword [rbp], rax
│  │╎││││   0x08001b40      call  fclose                               ; RELOC 32 fclose
│  │╎││││   ; CALL XREF from sym.numaWriteMem @ 0x8001b40
│  │╎││││   0x08001b45      add   rsp, 8
│  │╎││││   0x08001b49      mov   eax, r12d
│  │╎││││   0x08001b4c      pop   rbx
│  │╎││││   0x08001b4d      pop   rbp
│  │╎││││   0x08001b4e      pop   r12
│  │╎││││   0x08001b50      pop   r13
│  │╎││││   0x08001b52      ret
..
│  │╎││││   ; CODE XREF from sym.numaWriteMem @ 0x8001ae3
│  │╎││╰──> 0x08001b58      nop     dword [rax + rax]                  ; 5
│  │╎││╭──< 0x08001b5b      jle   0x8001bd0
│  │╎││││   ; CODE XREFS from sym.numaWriteMem @ 0x8001b87, 0x8001ba7, 0x8001beb
│ ╭───────> 0x08001b5d      add   rsp, 8
│ ╎│╎││││   0x08001b61      mov   r12d, 1
│ ╎│╎││││   0x08001b67      pop   rbx
│ ╎│╎││││   0x08001b68      mov   eax, r12d
│ ╎│╎││││   0x08001b6b      pop   rbp
│ ╎│╎││││   0x08001b6c      pop   r12
│ ╎│╎││││   0x08001b6e      pop   r13
│ ╎│╎││││   0x08001b70      ret
..
│ ╎│╎││││   ; CODE XREF from sym.numaWriteMem @ 0x8001ad3
│ ╎│╎│││╰─> 0x08001b78      nop     dword [rax]                        ; arg2
│ ╎│╎│││╭─< 0x08001b7b      je    0x8001b84
│ ╎│╎││││   0x08001b7d      mov   qword [rsi], 0                       ; arg2
│ ╎│╎││││   ; CODE XREF from sym.numaWriteMem @ 0x8001b7b
│ ╎│╎│││╰─> 0x08001b84      cmp   eax, 5                               ; 5
│ ────────< 0x08001b87      jg    0x8001b5d
│ ╎│╎│││    0x08001b89      mov   edx, 1
│ ╎│╎│││    0x08001b8e      lea   rsi, [0x08001b95]                    ; RELOC 32 .rodata @ 0x080032b0 - 0x8001a0d
│ ╎│╎│││    ; DATA XREF from sym.numaWriteMem @ 0x8001b8e
│ ╎│╎│││    0x08001b95      lea   rdi, [0x08001b9c]                    ; RELOC 32 .LC50 @ 0x08003039 - 0x8001b9c
│ ╎│╎│││    ; DATA XREF from sym.numaWriteMem @ 0x8001b95
│ ╎│╎│││╭─< 0x08001b9c      jmp   0x8001bbc
..
│ ╎│╎││││   ; CODE XREF from sym.numaWriteMem @ 0x8001b12
│ ╎╰──────> 0x08001ba0      nop                                        ; [0x8001ba7:4]=0x8d48b47f; RELOC 32 LeptMsgSeverity
│ ╎ ╎││││   ; DATA XREF from sym.numaWriteMem @ 0x8001ba0
│ ────────< 0x08001ba7      jg    0x8001b5d
│ ╎ ╎││││   0x08001ba9      lea   rsi, [0x08001bb0]                    ; RELOC 32 .rodata @ 0x080032b0 - 0x8001a28
│ ╎ ╎││││   ; DATA XREF from sym.numaWriteMem @ 0x8001ba9
│ ╎ ╎││││   0x08001bb0      lea   rdi, [0x08001bb7]                    ; RELOC 32 .LC53 @ 0x0800305d - 0x8001bb7
│ ╎ ╎││││   ; DATA XREF from sym.numaWriteMem @ 0x8001bb0
│ ╎ ╎││││   0x08001bb7      mov   edx, 1
│ ╎ ╎││││   ; CODE XREFS from sym.numaWriteMem @ 0x8001b9c, 0x8001be3, 0x8001c04
│ ─╭────╰─> 0x08001bbc      add   rsp, 8
│ ╎╎╎│││    0x08001bc0      pop   rbx
│ ╎╎╎│││    0x08001bc1      pop   rbp
│ ╎╎╎│││    0x08001bc2      pop   r12
│ ╎╎╎│││    0x08001bc4      pop   r13
│ ╎╎╎│││╭─< 0x08001bc6      jmp   returnErrorInt                       ; RELOC 32 returnErrorInt
│ ╎╎╎││││   ; CODE XREF from sym.numaWriteMem @ 0x8001bc6
│ ╎╎╎│││╰─> 0x08001bcb      nop   dword [rax + rax]
│ ╎╎╎│││    ; CODE XREF from sym.numaWriteMem @ 0x8001b5b
│ ╎╎╎││╰──> 0x08001bd0      mov   edx, 1
│ ╎╎╎││     0x08001bd5      lea   rsi, [0x08001bdc]                    ; RELOC 32 .rodata @ 0x080032b0 - 0x8001a54
│ ╎╎╎││     ; DATA XREF from sym.numaWriteMem @ 0x8001bd5
│ ╎╎╎││     0x08001bdc      lea   rdi, [0x08001be3]                    ; RELOC 32 .LC51 @ 0x0800304b - 0x8001be3
│ ╎╎╎││     ; DATA XREF from sym.numaWriteMem @ 0x8001bdc
│ ────────< 0x08001be3      jmp   0x8001bbc
..
│ ╎╎╎││     ; CODE XREF from sym.numaWriteMem @ 0x8001aef
│ ╎╎╎│╰───> 0x08001be8      nop     dword [rax]                        ; 5
│ ╰───────< 0x08001beb      jg    0x8001b5d
│  ╎╎│      0x08001bf1      mov   edx, 1
│  ╎╎│      0x08001bf6      lea   rsi, [0x08001bfd]                    ; RELOC 32 .rodata @ 0x080032b0 - 0x8001a75
│  ╎╎│      ; DATA XREF from sym.numaWriteMem @ 0x8001bf6
│  ╎╎│      0x08001bfd      lea   rdi, [0x08001c04]                    ; RELOC 32 .LC0 @ 0x08002d56 - 0x8001c04
│  │╎│      ; DATA XREF from sym.numaWriteMem @ 0x8001bfd
│  ╰──────< 0x08001c04      jmp   0x8001bbc
..
│   ╎│      ; CODE XREF from sym.numaWriteMem @ 0x8001b01
│   ╎╰────> 0x08001c10      nop     word cs:[rax + rax]                ; [0x8001c17:8]=0x48000000000d8d48; RELOC 32 stderr
│   ╎       ; DATA XREF from sym.numaWriteMem @ 0x8001c10
│   ╎       0x08001c17      lea   rcx, [0x08001c1e]                    ; RELOC 32 .rodata @ 0x080032b0 - 0x8001a96
│   ╎       ; DATA XREF from sym.numaWriteMem @ 0x8001c17
│   ╎       0x08001c1e      lea   rdx, [0x08001c25]                    ; RELOC 32 .LC52 @ 0x08003250 - 0x8001c25
│   ╎       ; DATA XREF from sym.numaWriteMem @ 0x8001c1e
│   ╎       0x08001c25      mov   esi, 1
│   ╎       0x08001c2a      xor   eax, eax
│   ╎       0x08001c2c      call  __fprintf_chk                        ; RELOC 32 __fprintf_chk
│   │       ; CALL XREF from sym.numaWriteMem @ 0x8001c2c
╰   ╰─────< 0x08001c31      jmp   tmpfile
[0x08001ac0]> pdg
AddressSanitizer:DEADLYSIGNAL
=================================================================
==1103556==ERROR: AddressSanitizer: SEGV on unknown address 0x000000000008 (pc 0x7f30e591b95c bp 0x7fff1a403a60 sp 0x7fff1a403a60 T0)
==1103556==The signal is caused by a READ memory access.
==1103556==Hint: address points to the zero page.
    #0 0x7f30e591b95c in AddrSpace::getType() const (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x4bd95c)
    #1 0x7f30e597fefe in Address::isConstant() const (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x521efe)
    #2 0x7f30e5975430 in Scope::stackCodeLabel(Scope const*, Scope const*, Address const&, LabSymbol**) (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x517430)
    #3 0x7f30e5975ea7 in Scope::queryCodeLabel(Address const&) const (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x517ea7)
    #4 0x7f30e5ab61a9 in PrintC::emitLabel(FlowBlock const*) (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x6581a9)
    #5 0x7f30e5ab1a60 in PrintC::emitGotoStatement(FlowBlock const*, FlowBlock const*, unsigned int) (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x653a60)
    #6 0x7f30e5ab44f7 in PrintC::emitBlockIf(BlockIf const*) (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x6564f7)
    #7 0x7f30e5ad8b27 in BlockIf::emit(PrintLanguage*) const (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x67ab27)
    #8 0x7f30e5ab3ba2 in PrintC::emitBlockGoto(BlockGoto const*) (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x655ba2)
    #9 0x7f30e5ad875f in BlockGoto::emit(PrintLanguage*) const (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x67a75f)
    #10 0x7f30e5ab4647 in PrintC::emitBlockIf(BlockIf const*) (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x656647)
    #11 0x7f30e5ad8b27 in BlockIf::emit(PrintLanguage*) const (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x67ab27)
    #12 0x7f30e5ab48d8 in PrintC::emitBlockIf(BlockIf const*) (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x6568d8)
    #13 0x7f30e5ad8b27 in BlockIf::emit(PrintLanguage*) const (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x67ab27)
    #14 0x7f30e5ab3e3e in PrintC::emitBlockLs(BlockList const*) (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x655e3e)
    #15 0x7f30e5ad89b9 in BlockList::emit(PrintLanguage*) const (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x67a9b9)
    #16 0x7f30e5ab3aab in PrintC::emitBlockGraph(BlockGraph const*) (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x655aab)
    #17 0x7f30e5ab3299 in PrintC::docFunction(Funcdata const*) (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x655299)
    #18 0x7f30e58e4160 in Decompile(rz_core_t*, unsigned long long, DecompileMode, std::__cxx11::basic_stringstream<char, std::char_traits<char>, std::allocator<char> >&, rz_annotated_code_t**) (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x486160)
    #19 0x7f30e58e463f in DecompileCmd(rz_core_t*, DecompileMode) (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x48663f)
    #20 0x7f30e58e5430 in pdg_handler(rz_core_t*, int, char const**) (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x487430)
    #21 0x7f30eb9d0b7c in argv_call_cb ../librz/core/cmd_api.c:618
    #22 0x7f30eb9d12ac in call_cd ../librz/core/cmd_api.c:651
    #23 0x7f30eb9d1290 in call_cd ../librz/core/cmd_api.c:648
    #24 0x7f30eb9d14fa in rz_cmd_call_parsed_args ../librz/core/cmd_api.c:669
    #25 0x7f30eb9a496d in handle_ts_arged_command_internal ../librz/core/cmd.c:4623
    #26 0x7f30eb9a3d4c in handle_ts_arged_command ../librz/core/cmd.c:4585
    #27 0x7f30eb9c0926 in handle_ts_command ../librz/core/cmd.c:6054
    #28 0x7f30eb9c1f8e in handle_ts_commands_internal ../librz/core/cmd.c:6111
    #29 0x7f30eb9c1155 in handle_ts_commands ../librz/core/cmd.c:6076
    #30 0x7f30eb9c3288 in core_cmd_tsr2cmd ../librz/core/cmd.c:6206
    #31 0x7f30eb9c3b51 in rz_core_cmd ../librz/core/cmd.c:6253
    #32 0x7f30eba15449 in rz_core_prompt_exec ../librz/core/core.c:2911
    #33 0x7f30eba12eb5 in rz_core_prompt_loop ../librz/core/core.c:2762
    #34 0x7f30f91c8045 in rz_main_rizin ../librz/main/rizin.c:1362
    #35 0x401a5c in main ../binrz/rizin/rizin.c:55
    #36 0x7f30f85f11e1 in __libc_start_main (/lib64/libc.so.6+0x281e1)
    #37 0x4011dd in _start (/home/akochkov/.local/bin/rizin+0x4011dd)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV (/home/akochkov/.local/share/rizin/plugins/core_ghidra.so+0x4bd95c) in AddrSpace::getType() const
==1103556==ABORTING
```

Binary is here: 
[numabasic.c-gcc-x64-O2.o.zip](https://github.com/rizinorg/rz-ghidra/files/6120409/numabasic.c-gcc-x64-O2.o.zip)

### PR DESCRIPTION
**Pull request by Crabtux**
Ghidra uses some scripts written in Java to fixup some analysis corner cases (see [Ghidra/Features/Base/ghidra_scripts](https://github.com/NationalSecurityAgency/ghidra/tree/master/Ghidra/Features/Base/ghidra_scripts)), but currently rz-ghidra is not able to use them, which causes some decompiling issues. This PR makes an attempt to fixup one of the corner cases. Details about the issue can be found in #202.

I try to solve the issue by adding a hook before decompiling action, but after the initialization of the RizinArchitecture. The hook function uses rizin to analyse shared return calls to relocs, and fixup the P-Code using `setFlowOverride()` API provided by Ghidra.

Expected to fix #202.

### CODE REVIEW

File: src/PcodeFixupPreprocessor.cpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+void PcodeFixupPreprocessor::fixupSharedReturnJumpToRelocs(RzAnalysisFunction *function, Funcdata *func, RzCore *core, RizinArchitecture &arch)
+{
+	RzList *refs = rz_analysis_function_get_xrefs_from(function);
+
+	RzListIter *it;
+	RzAnalysisXRef *xref;
+
+	// C++ has more strict type checking than C, which stops us from implicitly casting void * to RzListIter *.
+	// Expand the rz_list_foreach macro manually. 
+	if (refs) for (it = refs->head; it && (xref = (RzAnalysisXRef *)it->elem, 1); it = it->next)
```
Comment (wargio) [inline src/PcodeFixupPreprocessor.cpp]: potentially you could reuse (and rename) https://github.com/rizinorg/cutter/blob/dev/src/core/RizinCpp.h#L110 or https://github.com/rizinorg/cutter/blob/dev/src/core/RizinCpp.h#L56

Comment (Crabtux) [reply src/PcodeFixupPreprocessor.cpp]: It seems that rz-ghidra has already implemented `rz_list_foreach_cpp` in [RizinUtils.h](https://github.com/rizinorg/rz-ghidra/blob/dev/src/RizinUtils.h). Maybe I can rewrite my code to use that :)

**Comment by XVilka** [comment]
The red CI should be fixed once https://github.com/rizinorg/cutter/pull/3319 is merged

**Review by XVilka — APPROVED** [review]
@thestr4ng3r take a look please

**Comment by thestr4ng3r** [comment]
Thanks! The test case is also very useful.

**Pull request merged**